src/monzo.ts

```diff
@@ -444,3 +370,203 @@ function bigIntPrimeLimit(
+  [45, M8], [7, M3], [1, M0], [3, M1], [1, M0], [5, M2], [3, M1], [1, M0], [7, M3], [9, M4], [5, M2], [1, M0], [3, M1], [1, M0], [1, M0],
+  [105, Ma], [1, M0], [1, M0], [9, M4], [1, M0], [5, M2], [3, M1], [7, M3], [1, M0], [3, M1], [5, M2], [1, M0], [9, M4], [1, M0], [7, M3],
+  [15, M5], [1, M0], [1, M0], [3, M1], [1, M0], [5, M2], [63, M9], [1, M0], [1, M0], [3, M1], [5, M2], [1, M0], [3, M1], [7, M3], [1, M0],
+  [45, M8], [1, M0], [1, M0], [3, M1], [1, M0], [35, M7], [3, M1], [1, M0], [1, M0], [9, M4], [5, M2], [1, M0], [21, M6], [1, M0], [1, M0],
+  [15, M5], [1, M0], [1, M0], [9, M4], [7, M3], [5, M2], [3, M1], [1, M0], [1, M0], [3, M1], [5, M2], [7, M3], [9, M4], [1, M0], [1, M0],
+  [15, M5], [1, M0], [1, M0], [21, M6], [1, M0], [5, M2], [9, M4], [1, M0], [1, M0], [3, M1], [35, M7], [1, M0], [3, M1], [1, M0], [1, M0],
+  [45, M8], [1, M0], [7, M3], [3, M1], [1, M0], [5, M2], [3, M1], [1, M0], [1, M0], [63, M9], [5, M2], [1, M0], [3, M1], [1, M0], [1, M0],
+  [15, M5], [7, M3], [1, M0], [9, M4], [1, M0], [5, M2], [3, M1], [1, M0], [7, M3], [3, M1], [5, M2], [1, M0], [9, M4], [1, M0], [1, M0],
+  [105, Ma], [1, M0], [1, M0], [3, M1], [1, M0], [5, M2], [9, M4], [7, M3], [1, M0], [3, M1], [5, M2], [1, M0], [3, M1], [1, M0], [7, M3],
+  [45, M8], [1, M0], [1, M0], [3, M1], [1, M0], [5, M2], [21, M6], [1, M0], [1, M0], [9, M4], [5, M2], [1, M0], [3, M1], [7, M3], [1, M0],
+  [15, M5], [1, M0], [1, M0], [9, M4], [1, M0], [35, M7], [3, M1], [1, M0], [1, M0], [3, M1], [5, M2], [1, M0], [63, M9], [1, M0], [1, M0],
+  [15, M5], [1, M0], [1, M0], [3, M1], [7, M3], [5, M2], [9, M4], [1, M0], [1, M0], [3, M1], [5, M2], [7, M3], [3, M1], [1, M0], [1, M0],
+  [45, M8], [1, M0], [1, M0], [21, M6], [1, M0], [5, M2], [3, M1], [1, M0], [1, M0], [9, M4], [35, M7], [1, M0], [3, M1], [1, M0], [1, M0],
+  [15, M5], [1, M0], [7, M3], [9, M4], [1, M0], [5, M2], [3, M1], [1, M0], [1, M0], [21, M6], [5, M2], [1, M0], [9, M4], [1, M0], [1, M0],
+  [15, M5], [7, M3], [1, M0], [3, M1], [1, M0], [5, M2], [9, M4], [1, M0], [7, M3], [3, M1], [5, M2], [1, M0], [3, M1], [1, M0], [1, M0]
+];
+
+const BIG_LOOKUP: [bigint, Monzo][] = MONZO_LOOKUP.map(l => [
+  BigInt(l[0]),
+  l[1],
+]);
+
+// XXX: Not resistant to zero, check your inputs.
+function intToMonzo7(n: number): [Monzo, number] {
+  const result = [0, 0, 0, 0];
+  // Factors of two using bit magic
+  while (!(n & 1)) {
+    const twos = (n ^ (n - 1)) & n;
+    n /= twos;
+    result[0] += Math.log2(twos);
+  }
+  while (true) {
+    const [factor, m] = MONZO_LOOKUP[n % 315];
+    if (factor === 1) {
+      break;
+    }
+    n /= factor;
+    result[1] += m[1];
+    result[2] += m[2];
+    result[3] += m[3];
+  }
+  return [result, n];
+}
+
+// XXX: Not resistant to zero, check your inputs.
+function bigIntToMonzo7(n: bigint): [Monzo, bigint] {
+  const result = [0, 0, 0, 0];
+  // Factors of two using bit magic
+  const twos = (n ^ (n - 1n)) & n;
+  n /= twos;
+  result[0] += Math.log2(Number(twos));
```

Comment:
I guess this is fast but right near bit magic it feels weird. Though of course bit magic is here just to count zeros at least semi-efficiently and not to be blazing fast, so there’s no contradiction anywhere in the end.

Reply:
This makes me uneasy as well. I'll add an apology in a comment.

Reply:
Actually breaks with `2n ** 1024n`. :sweat_smile: